Added Errorstats info section to enable keeping track of the different errors that occur within Redis; Added failed_calls to commandstats

filipecosta90 opened the pull request:
this PR pushes forward the observability on overall error statistics and command statistics within redis-server:

- It extends `INFO COMMANDSTATS` to have
  - failed_calls in - so we can keep track of errors that happen from the command itself, broken by command.
  - rejected_calls - so we can keep track of errors that were triggered outside the commmand processing per se ( example of wrong arity, oom, etc... )
Sample ( notice rejected_calls ):
  ```
  127.0.0.1:6379> set k 
  (error) ERR wrong number of arguments for 'set' command
  127.0.0.1:6379> info commandstats
  # Commandstats
  cmdstat_set:calls=0,usec=0,usec_per_call=0.00,rejected_calls=1,failed_calls=0
  cmdstat_info:calls=8,usec=1077,usec_per_call=134.62,rejected_calls=0,failed_calls=0
  ```

- Adds a new section to INFO, named `ERRORSTATS` that enables keeping track of the different errors that occur within redis ( within processCommand and call ) based uppong the reply Error Prefix ( The first word after the "-", up to the first space ). 
  
  Sample:
  ```
  127.0.0.1:6379> auth a
  (error) ERR AUTH <password> called without any password configured for the default user. Are you sure your configuration is correct?
  127.0.0.1:6379> set k
  (error) ERR wrong number of arguments for 'set' command
  127.0.0.1:6379> ACL SETUSER alice on >p1pp0 ~cached:* +get +info
  OK
  127.0.0.1:6379> auth alice p1pp0
  OK
  127.0.0.1:6379> set a b
  (error) NOPERM this user has no permissions to run the 'set' command or its subcommand
  127.0.0.1:6379> info errorstats
  # Errorstats
  errorstat_ERR:count=2
  errorstat_NOPERM:count=1
  ```


- This PR also fixes RM_ReplyWithError so that it can be correctly identified as an error reply.

---
To test the added features:
```
tclsh tests/test_helper.tcl --verbose --single unit/info
tclsh tests/cluster/run.tcl --single tests/18-info.tcl
```

The added tests are divided as follow:
- **failed** call authentication error
- **failed** call NOGROUP error
- **rejected** call due to wrong arity
- **rejected** call by OOM error
- **rejected** call by authorization error
- **rejected** call rejected due to MOVED Redirection

oranagra commented on src/networking.c:
```diff
@@ -399,6 +399,25 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
 void afterErrorReply(client *c, const char *s, size_t len) {
+    struct redisError *err = NULL;
+    /* Errors that occurred during command processing */
+    if (c->cmd){
```
```suggestion
    if (c->cmd) {
```

oranagra commented on src/networking.c:
```diff
@@ -399,6 +399,25 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+        err = lookupError(shared.authorizationerror);
+        if (err) err->count++;
+    /* Authentication errors */
+    } else if (!strncmp(s, "AUTH", 4) ||
+               !strncmp(s, "-NOAUTH", 7) ||
+               !strncmp(s, "-WRONGPASS", 10)){
```
why isn't `AUTH` (the first error) prefixed with `-`?
```suggestion
    } else if (!strncmp(s, "AUTH", 4) ||
               !strncmp(s, "-NOAUTH", 7) ||
               !strncmp(s, "-WRONGPASS", 10))
    {
```

madolson replied:
Can we also just move the pointer one character in, we know the first character must be "-"

madolson replied:
Also, isn't this duplicative of the ACL log?

filipecosta90 replied:
> why isn't `AUTH` (the first error) prefixed with `-`?

@oranagra due to "AUTH" error string not starting with `-`. The `-` is added afterwards but non included in the error string, within addReplyErrorLength() line `if (!len || s[0] != '-') addReplyProto(c,"-ERR ",5);` .
Noneteless if we follow the new approach this is accounted for...

oranagra replied:
ok, this means that the text "AUTH" which we matched against, is not the error code, it's just a part of the error text which happens to be uppercase.
i think it means we need to move `afterErrorReply` to be called in the different place, since it takes an error message, but it's no the final error message.

oranagra commented on src/server.c:
```diff
@@ -183,885 +183,891 @@ struct redisServer server; /* Server global state */
     {"module",moduleCommand,-2,
      "admin no-script",
-     0,NULL,0,0,0,0,0,0},
+     0,NULL,0,0,0,0,0,0,0},
```
i would like to try to find a way to avoid there changes (gonna cause conflicts for other PRs).

one way is to group `microseconds`, `calls`, `failed_calls`, and `id` into a separate struct for runtime info. then we can init that entire struct with one `{}`, and we don't need to edit the declarative statement whenever a runtime variable is added.

but the above idea still means we need to edit it once (today),
let's try to think of a way..

filipecosta90 replied:
Agree :) Will do that change - since yesterday we already had two conflicts =) )

oranagra replied:
we have a plan for the future, but i'm also hoping to find a plan for now (would very much like to avoid these modifications)

oranagra replied:
I think we can just omit that change for now, and it'll all pass.
it means that the last `0` which was the command `id` will now be stored into the new `failed_calls` and the last member (the `id`) will be implicitly initialized to 0 (which is ok).

i remember in the past that some compiler generated a warning on structs that were partially initialized but not fully initialized, but i see in the CI for `#8210` passes and we do use `-Werror`.
ideally that means we can trim the last 3 `0`, and not just avoid adding a 4th, but i don't wanna do that.

i'm planning another huge PR soon which will completely rewrite that whole command table array, so that's when i'll trim these.

filipecosta90 replied:
@oranagra followed the suggestion and removed this changes ( it makes the PR a lot cleaner )

oranagra commented on src/server.c:
```diff
@@ -183,885 +183,891 @@ struct redisServer server; /* Server global state */
+struct redisError redisErrorTable[] = {
+    {"authorization",0}, /* Anything blocked by ACL */
+    {"authentication",0}, /* Anything blocked by missing/wrong authentication */
+    {"command",0},  /* Errors from valid commands that failed (e.g. missing key) */
```
i don't see why we need this static struct and a fancy dict population function.
at the end of which, we have the error name in both the dict key, and again embedded inside it's value.

one other way is to drop this static table, and just have a 3 line population function that adds the 3 errors (3 calls to dictAdd), later using dictGetUnsignedIntegerVal to increment.
or is we need a struct with more than one integer, we can calloc it when calling dictAdd.

The way i see it, the justification for the command table population way, is when you have a lot of metadata to add to each item on creation, and you want it shown in the code in a "declarative" form. but in this case we don't have a metadata (it's initialized with 0)

another alternative is to not use dict at all, just an array with some enum as index to it.
the fact is, that currently, you're not looking up for an error that was given to you by the caller (which justifies a runtime lookup), you're only actually looking up an error which you know (at compile time) which one it is, so you can just use an enum as an index to the array.
this will not hold if we'll try to detect other types of errors in the future, in which case we'll need `lookupError` to only match the error prefix, and not the entire string.

filipecosta90 replied:
@oranagra followed your advice and using a rax with a much simpler and future proof aproach. can you check it?

oranagra commented on src/server.c:
```diff
@@ -2386,6 +2402,13 @@ void createSharedObjects(void) {
     shared.queued = createObject(OBJ_STRING,sdsnew("+QUEUED\r\n"));
     shared.emptyscan = createObject(OBJ_STRING,sdsnew("*2\r\n$1\r\n0\r\n*0\r\n"));
+
+    /* The following shared objects, are related to error categories */
+    shared.commanderror = sdsnew("command");
+    shared.authorizationerror = sdsnew("authorization");
+    shared.authenticationerror = sdsnew("authentication");
```
in the current form, these are not needed, the lookup dict (if at all we keep it), can use a char* search rather than an sds search, and the char* can be a static string, not a heap allocation.

filipecosta90 replied:
fixed in last commit

oranagra commented on src/networking.c:
```diff
@@ -399,6 +399,25 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+        if (err) err->count++;
+    }
+    /* Authorization error counter (commands blocked by ACLs) */
+    if (!strncmp(s, "-NOPERM", 7)){
```
```suggestion
    if (!strncmp(s, "-NOPERM", 7)) {
```

oranagra reviewed (commented):
@filipecosta90 thanks for this PR.
i added a few comments about the implementation, but maybe we should first discuss the key features here.

as we've established, there are generally two groups of errors, ones that are handled in processCommand and ones that happen inside the command code itself.
do we intend to cover only the first group? or also the second?
If we intend to extend this feature to match other types of errors, it's probably wrong to do an if-else chain of calls to `strncmp`, and then a lookup in a dict (or an enum lookup into an array).
maybe we should have a rax in which we can search by prefix, and implicitly support all type of errors.

i think it would be nice to examine this feature after adding a few more errors as an example (like OOM, WRONGTYPE, NOSCRIPT, LOADING, BUSY, BUSYKEY.
one thing we'll soon discover is that many of these are being used with a simple addReply (so we don't even have them on our radar).

madolson commented on src/networking.c:
```diff
@@ -399,6 +399,25 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+    } else if (!strncmp(s, "AUTH", 4) ||
+               !strncmp(s, "-NOAUTH", 7) ||
+               !strncmp(s, "-WRONGPASS", 10)){
+        err = lookupError(shared.authenticationerror);
```
Can this function just increment the error? Not sure why we need to fetch it then increment it.

filipecosta90 replied:
agree @madolson . changed in the latest commit

madolson commented:
Also related to Oran's top level comment, we don't always use "addReplyError*" for all of the errors, we need to go through and audit them.

filipecosta90 commented:
> do we intend to cover only the first group? or also the second?
I believe we can see value out of both groups right? meaning the more errors we can keep track without paying a penalty the better.

> If we intend to extend this feature to match other types of errors, it's probably wrong to do an if-else chain of calls to `strncmp`, and then a lookup in a dict (or an enum lookup into an array).
> maybe we should have a rax in which we can search by prefix, and implicitly support all type of errors.
> i think it would be nice to examine this feature after adding a few more errors as an example (like OOM, WRONGTYPE, NOSCRIPT, LOADING, BUSY, BUSYKEY.

@oranagra @madolson, following what you've mentioned I've updated the code to use the Error Prefix ( The first word after the "-", up to the first space ) as the rax key (lowercased). In that manner we have the desired implicit support for all errors. Can you guys check the new commit. 

Sample output of errorstats now:
```
127.0.0.1:6379> auth a
(error) ERR AUTH <password> called without any password configured for the default user. Are you sure your configuration is correct?
127.0.0.1:6379> set k
(error) ERR wrong number of arguments for 'set' command
127.0.0.1:6379> info errorstats
# Errorstats
errorstat_auth:count=1
errorstat_wrong:count=1
```

oranagra commented:
@filipecosta90 looking at the example in your last post (didn't read the code yet), i think these two cases should fall into the standard `err` since the error code is `-ERR` ("wrong" and "AUTH" are just part of the text).  
i.e.
```
errorstat_err:count=2
``` 
not
```
errorstat_auth:count=1
errorstat_wrong:count=1
```

if we think it is important for redis to distinguish between them, then it's also important for client libraries, and we should change the code and promote them to have their own error code.

oranagra commented on src/networking.c:
```diff
@@ -399,6 +399,20 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+    char* spaceloc = strchr(s, ' ');
+    size_t padding = 0;
+    // remove spurious dash
+    if (s[0] == '-') padding++;
```
we do not want to **optionally** remove the spurious dash, this is in fact an indication that this method (`afterErrorReply`) is called in the wrong place, and given the wrong input (incomplete error message).
if we do that, the dash will always be present (and we can remove it, and even assert that it's there).

filipecosta90 replied:
agree. fixed in the latest commit and squash. please see my latest comment

oranagra commented on src/server.c:
```diff
@@ -3292,6 +3295,17 @@ void resetCommandTableStats(void) {
 }
 
+void resetErrorTableStats(void) {
+    raxIterator ri;
+    raxStart(&ri,server.errors);
+    raxSeek(&ri,"^",NULL,0);
+    while(raxNext(&ri)) {
+        struct redisError *err = ri.data;
+        if(err) err->count=0;
+    }
+    raxStop(&ri);
```
i think you just need to discard the rax and create a fresh one.
this way the INFO response won't even show 0 values for errors that no longer happen.

filipecosta90 replied:
@oranagra given that we check on the `info errorstats` string building if the error is 0 and don't print it (`if(e->count)`) shouldn't we keep the rax to avoid uncessary destroy and construction?

Sample output validating the above:
```
127.0.0.1:6379> auth 1
(error) ERR AUTH <password> called without any password configured for the default user. Are you sure your configuration is correct?
127.0.0.1:6379> info errorstats
# Errorstats
errorstat_auth:count=1
127.0.0.1:6379> config resetstat
OK
127.0.0.1:6379> info errorstats
# Errorstats
```

oranagra replied:
yeah, you're right about the output. sorry,
well, many places in redis just discard the rax or dict and create an empty one in similar situation, i suppose it avoids keeping a lot of elements that are relics past traffic (in this case errors might no longer be seen again).
but i suppose there are not that may error types, so i guess we can go either way.

oranagra commented on src/server.c:
```diff
@@ -3952,6 +3966,22 @@ int processCommand(client *c) {
+    struct redisError *error = raxFind(server.errors,(unsigned char*)fullerr,namelen);
+    if (error == raxNotFound) {
+        error = zmalloc(sizeof(*error));
+        error->name = zmalloc(namelen+1);
+        memcpy(error->name,fullerr,namelen);
+        error->name[namelen] = '\0';
```
you don't need to store the name inside the rax "value".
when iterating on the rax you get the key from the iterator.
considering my previous comments, the error code will always be uppercase. no need to mess with the character casing.

filipecosta90 replied:
fixed in the last commit and squash

filipecosta90 commented:
> we do not want to **optionally** remove the spurious dash, this is in fact an indication that this method (`afterErrorReply`) is called in the wrong place, and given the wrong input (incomplete error message).
> if we do that, the dash will always be present (and we can remove it, and even assert that it's there).

@oranagra agree. In the latest code push I assume that if no `-` is present than the error is the default `ERR` (same as done in `addReplyErrorLength`), otherwise we use the first after the `-` until the first space. 

I've also removed the error name from the error struct and I'm using the error prefix as is ( meaning no lowercase ). 

I believe all comments have been address. wdyt?

oranagra commented on src/networking.c:
```diff
@@ -405,6 +405,22 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+     * If the string already starts with "-..." then the error prefix
+     * is provided by the caller. Otherwise we use "-ERR". */
+    if (s[0] != '-') incrementError("ERR", 3);
+    else {
```
as much as i hate wasting lines on braces, this seems to be the style in redis
```suggestion
    if (s[0] != '-') {
        incrementError("ERR", 3);
    } else {
```

oranagra commented on src/networking.c:
```diff
@@ -405,6 +405,22 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+     * is provided by the caller. Otherwise we use "-ERR". */
+    if (s[0] != '-') incrementError("ERR", 3);
+    else {
+        char* spaceloc = strchr(s, ' ');
```
we need to limit this search to `len`, but maybe also lower?
like `MIN(len,32)`?

filipecosta90 replied:
fixed in the last commit ( added description to ensure who's reading the code know the max error prefix length used is 32

oranagra commented on src/networking.c:
```diff
@@ -405,6 +405,22 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+    if (s[0] != '-') incrementError("ERR", 3);
+    else {
+        char* spaceloc = strchr(s, ' ');
+        if (spaceloc) {
```
i suppose we need to have an `else`, either an assertion, or falling back to add `ERR` type error.
i think I would have been ok with an assertion, but then if i think of modules, maybe that's not wise.

filipecosta90 replied:
agree. fixed in the last commit

oranagra commented on src/server.c:
```diff
@@ -3287,11 +3288,23 @@ void resetCommandTableStats(void) {
+    raxSeek(&ri,"^",NULL,0);
+    while(raxNext(&ri)) {
+        struct redisError *err = ri.data;
+        if(err) err->count=0;
```
i think we never expect this to be NULL, let i rather drop the `if` and let it crash if we're wrong.
```suggestion
        err->count=0;
```

oranagra commented on src/server.c:
```diff
@@ -4903,14 +4928,32 @@ sds genRedisInfoString(const char *section) {
         dictReleaseIterator(di);
     }
+    /* Error statistics */
+    if (allsections || !strcasecmp(section,"errorstats")) {
```
just noting that this section is not included by default. i guess that's good.

filipecosta90 replied:
@oranagra given this is so "cheap" to compute and potentially a very usefull info I've added it as default. wdyt?

oranagra replied:
i see the only section that's currently not part of the default is the commandstats, i don't suppose it's a matter of being cheap to generate (other parts are possibly more demanding), i think it's also about the verbosity and usefulness.
i suppose that in practice we'll never see here more than some 4 errors, but i'm not sure i'd say it's more useful than the commands stats.

let's leave this question open for later discussion.

oranagra commented on src/server.c:
```diff
@@ -4903,14 +4928,32 @@ sds genRedisInfoString(const char *section) {
+        struct redisError *e;
+        while(raxNext(&ri)) {
+            e = (struct redisError *) ri.data;
+            if(e->count)
```
let's replace that with `if (!c->count) continue;` so that we don't have the extra indentation.

styling: both the `if` and `while` above are missing a space (please search the rest of your (new) code and fix these (maybe grepping the diff file).

other personal taste styling:
* i don't like a multi-line action in an `if` without braces.
* i don't like statements that change execution flow (break, continue, return, goto) being in the same line together with the `if`. unless this is an error handling code that is heavily repeated, in which case i rather not waste too many lines on it, reducing the amount of the related code that fits in the screen.

filipecosta90 replied:
fixed in the last commit. thank you for the explanation Oran.

oranagra commented on tests/unit/info.tcl:
```diff
@@ -0,0 +1,58 @@
+start_server {tags {"info"}} {
+    start_server {} {
+        set master_host [srv 0 host]
+        set master_port [srv 0 port]
```
these are not needed / used
```suggestion
```

oranagra commented on tests/unit/info.tcl:
```diff
@@ -0,0 +1,58 @@
+            assert_match {} [errorstat ERR]
+        }
+
+        test {errorstats: OOM error} {
```
let's add a test that checks `-NOGROUP` and `-WRONGTYPE`, and an invalid command like `catch {r asdfasdf}`.
both are different classes of errors than the ones you tested.
i.e. one will have both `calls` and `failed_calls` showing `1`. and the other will show an error with nothing in the command table.

This attempt will immediately expose two facts:
1. some commands fail with an error without being executed, and others fail inside the command. maybe we want to distinguish between `failed_calls` and `rejected_calls` in the command stats? (we can do that in `call()`)

2. the problem which was discussed before, which is that many commands use a plain `addReply` to reply with an error, and we can't log these today.
the only way out that i can think of is to grep for `addReply(*,shared.*err);` and modify each of these to use a new `addErrorReply` or `addReplyErrObject`.
and even that's not enough since i see some places do funny things like:
```c
    addReplySds(c,
        sdsnew("-WRONGTYPE Key is not a valid "
               "HyperLogLog string value.\r\n"));
```
if we do this, let's do that in a separate commit, so a reviewer (now or in the future) on your new code won't have to scan changes in all other files. (i.e. we won't squash-merge it)

filipecosta90 replied:
@oranagra with regards to the tests I've added the requested ones and now we have:
The added tests are divided as follow:
- **failed** call authentication error
- **failed** call NOGROUP error
- **rejected** call due to wrong arity
- **rejected** call by OOM error
- **rejected** call by authorization error

With regards to 1) accounting for rejected vs failure I've added a new flag to the command when we call `rejectCommand` or `rejectCommandFormat`, meaning if we reject a command it wont increment the failed counter. 

With regards to 2) I believe @madolson has already kicked off that effort on https://github.com/redis/redis/pull/8226 so we should push 2) there correct?

I believe all comments have been addressed. Can you check if the added changes make sense? Thank you both for reviewing this and pushing this feature forward.

oranagra replied:
sorry, i'm not able to catch up on my notifications lately, so i wasn't aware of that other PR.
i didn't review all your changes yet, but i do wee you addressed this comment.
the only things in this comment that are not addressed are:
* WRONGTYPE, which you can't address yet, until the other PR is merged, but also that's the reason i suggested the NOGROUP error, since it's already in the state today that the WRONGTYPE will be after the other PR is merged. so i guess we can skip that (or after the other PR is merged, replace the uncommon NOGROUP test with a simpler common WRONGTYPE test)
*  the unknown (asdf) command (which will leave a trace in the error stats, and not in the command stats). i think we should add it.

filipecosta90 replied:
@oranagra with regards to the unknown command test just added it in the latest commit.

madolson commented on src/server.c:
```diff
@@ -3287,11 +3288,24 @@ void resetCommandTableStats(void) {
 
 }
 
+void resetErrorTableStats(void) {
```
This is slightly weird user behavior, since reset doesn't actually reset the table back to what it looked like at the start. I would free the table and create a new one, and set the explicit expectation that if an element is missing, it's never been seen.

madolson commented on src/server.h:
```diff
@@ -1580,14 +1584,18 @@ struct redisCommand {
                    bit set in the bitmap of allowed commands. */
 };
 
+struct redisError {
+    long long count;
+};
```
Do we intend to extend this? Otherwise it would be easier to just store the count in the rax.

filipecosta90 replied:
@madolson I thought of it as wrapper struct for future requirements.

madolson replied:
Okay, as long as that is intended, it's okay.

madolson commented on src/networking.c:
```diff
@@ -405,6 +405,28 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+    /* Update per command error stats.
+     * If the command was rejected no need to flag it
+     * as a failed count also. */
+    if (c->cmd && !(c->cmd->flags & CMD_ERR_REJECTED)) {
```
where does CMD_ERR_REJECTED get reset?

filipecosta90 replied:
you're totally right @madolson. I've fixed it and added a test to confirm it's working as expected

oranagra commented:
@filipecosta90 it looks like you added a bunch of commits, and a rebase/merge from unstable, and then squashed it all, and i can't seem to be able to review the changes easily.
please avoid it in the future.

i'm ok with amending a commit and force pushing it (github let's me view the diffs), or to add more and more commits which will then be squashed when the PR is merged.
rebase from unstable or squashing your commits should be done when no one is still looking at the PR, and if you have to grab fresh code from unstable, do it with merge and no squashing, so i can skip that commit.

oranagra commented on src/networking.c:
```diff
@@ -405,6 +405,32 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+    /* Update per command error stats.
+     * If the command was rejected no need to flag it
+     * as a failed count also. */
+    if (c->cmd) {
+        if (c->cmd->flags & CMD_ERR_REJECTED) {
+            c->cmd->rejected_calls++;
+        } else {
+            c->cmd->failed_calls++;
+        }
+    }
```
i think this function should only update the error stats, and not the command stats.
in theory a single command can reply with an array of errors (more than one error per call).

i think the right place to update the command stats, is at the end of `call()` and in `rejectCommand()`, both are nontenured to be executed only once per command execution.

this means that the new `CMD_ERR_REJECTED` flag you added isn't needed.
and instead you do need to be able to tell in `call()` that the command failed.
the way to do that can be to try to match a global error counter (much like it looks at changes in `server.dirty`). this new global error counter can also be useful as a general metric (put next to `stat_unexpected_error_replies`)

the only problem with this is nested command (Lua, Multi-Exec, and Modules).
we can probably solve it by adding (yet another server struct global) which we reset to the last global error counter, every time we "log" the error into some command.

filipecosta90 replied:
@oranagra @madolson I've followed the review suggestion and removed the added flag CMD_ERR_REJECTED and moved all command error log to processCommand+rejectCommand. Oran I'm checking within processCommand and not call so that we can check the initial error count and then after call or queing the command we can see if any error reply occurred. Given that rejectCommand/rejectCommandFormat return prior than call this will simplifies stuff and allows to differentiate rejected from failed. 
what do you guys think of this approach?

madolson commented:
@oranagra Did you look at the code where we are temporarily attaching a flag to a command? That doesn't seem like a very maintainable assumption to make.

oranagra commented:
@madolson i agree. i'm still editing my CR post when i'm suggesting a better alternative.

oranagra commented on src/networking.c:
```diff
@@ -405,6 +405,32 @@ void addReplyErrorLength(client *c, const char *s, size_t len) {
+        if (spaceloc) {
+            const size_t errEndPos = (size_t)(spaceloc - s);
+            incrementError(s+1, errEndPos-1);
+        /* Fallback to ERR if we can retrieve the error prefix */
+        } else {
```
```suggestion
        } else {
            /* Fallback to ERR if we can retrieve the error prefix */
```

oranagra commented on src/server.h:
```diff
@@ -1236,6 +1233,7 @@ struct redisServer {
     size_t stat_module_cow_bytes;   /* Copy on write bytes during module fork. */
     uint64_t stat_clients_type_memory[CLIENT_TYPE_COUNT];/* Mem usage by type */
     long long stat_unexpected_error_replies; /* Number of unexpected (aof-loading, replica to master, etc.) error replies */
+    long long stat_total_error_replies; /* Total number of issued error replies ( command + rejected errors ) */
```
let's add it to the INFO STATS section

filipecosta90 replied:
added in the last commit

oranagra commented on src/server.c:
```diff
@@ -3974,6 +3971,11 @@ int processCommand(client *c) {
             handleClientsBlockedOnKeys();
     }
+
+    /* Update failed command calls if required */
+    if (previousErrCount < server.stat_total_error_replies) {
+        c->cmd->failed_calls++;
+    }
```
this may need to be in `call()` not `processCommand()`.
let's try to think this though together.
when someone does:
```
MULTI
SET foo bar
LPUSH foo bla
EXEC
```
do we want the failure to be logged in LPUSH, EXEC, or both?
I think we only want it in LPUSH, so i think we need to move this block to `call`, but also add another global variable that will "consume" this diff in some way, so that EXEC will not show that error.
similar thing will happen for Lua and Modules doing RM_Call.

filipecosta90 replied:
@oranagra I've added two tests for multi and lua to showcase that with the latest push we're properly setting the error on the right command and not having it shown on exec/eval,etc... 
Tests: https://github.com/redis/redis/pull/8217/files#diff-595bcf8dbabe8679a083b9f9480389d8526888760af5de2a6941c715c2a00d82R23
I'm basically checking if we're within a multi context or lua or module client and if so I'm restoring the error back to the original previous value given the failed counter it has already been updated on the real command as follow:
```c
const long long prev_err_count = server.stat_total_error_replies;
    updateCachedTime(0);
    start = server.ustime;
    c->cmd->proc(c);
    duration = ustime()-start;
    dirty = server.dirty-dirty;
    if (dirty < 0) dirty = 0;

    /* Update failed command calls if required */
    if ((server.stat_total_error_replies - prev_err_count)>0) {
        real_cmd->failed_calls++;
        /* If we're within a MULTI context or LUA or RM_Call
         * we don't propagate the error uppwards */
        if ( c->flags & CLIENT_MULTI ||
             c->flags & CLIENT_LUA ||
             c->flags & CLIENT_MODULE ) {
            server.stat_total_error_replies=prev_err_count;
        }
    }
```
what do you guys think?

oranagra replied:
@filipecosta90 i had something else in mind. 
Note that the way you implemented it, the errors inside the multi will not be shown in the global error counter. 

What I thought of doing is store another global variable (possibly a static inside the function), which recalls the last value of the global error counter which was used to increment the command error counter, so that one increment of the global error counter can't increment a command counter of more than once..

maybe it's even enough to make the stack variable you added in `call` a static one. Then, we update it in the entrance to `call`, and check again it at exit (before we update it again). 
This way we never set back the global error counter, but we also never log the same error twice in nested commands.

filipecosta90 replied:
@oranagra I've followed you recommendation with regards to the static variable within `call` and added an extra test (rejected call within multi/exec). 
Apart from that also checking the server stat total_error_replies value. 
Can you check it?

oranagra commented on tests/support/util.tcl:
```diff
@@ -555,12 +555,24 @@ proc get_child_pid {idx} {
     return $child_pid
 }
 
+proc statsrstat {key r} {
+    if {[regexp "\r\n$key:(.*?)\r\n" [$r info stats] _ value]} {
+        set _ $value
+    }
+}
+
```
there's a method called `s` which does this:
```
[s total_error_replies]
```

filipecosta90 replied:
thank you @oranagra, addressed in the last commit

oranagra commented on src/server.c:
```diff
@@ -3507,13 +3517,19 @@ void call(client *c, int flags) {
     dirty = server.dirty-dirty;
     if (dirty < 0) dirty = 0;
 
+    /* Update failed command calls if required */
+    if ((server.stat_total_error_replies - prev_err_count)>0) {
```
i think we better add a comment somewhere explaining the use of the static variable.
i.e. that it is used so that nested calls to `call` don't log the same error twice.
here is probably a good place.

filipecosta90 replied:
@oranagra added the details on why we're required the static variable. please revise

oranagra commented:
@redis/core-team please approve new info fields:
a per command stat `rejected_calls` and `failed_calls`
a per-error type counters (new info section)
and a new stat for total error count.
details at the top.

@filipecosta90 can you please make a redis-doc PR? thank you!

filipecosta90 commented:
> @filipecosta90 can you please make a redis-doc PR? thank you!

@oranagra @itamarhaber with regards to the redis-doc PR can you guys check https://github.com/redis/redis-doc/pull/1472 ?

madolson commented on src/server.c:
```diff
@@ -4904,14 +4941,34 @@ sds genRedisInfoString(const char *section) {
         while((de = dictNext(di)) != NULL) {
             c = (struct redisCommand *) dictGetVal(de);
-            if (!c->calls) continue;
+            if (!c->calls && !c->failed_calls &&!c->rejected_calls) 
```
style: space between && and !c->rejected_calls

madolson commented on src/server.c:
```diff
@@ -3958,9 +3980,22 @@ int processCommand(client *c) {
+        error->count = 0;
+        raxInsert(server.errors,(unsigned char*)fullerr,namelen,error,NULL);
+    }
+    if (error) error->count++;
```
When can error ever be NULL? I don't see how it's possible, and it would be nice if we could remove it.

madolson commented on src/server.c:
```diff
@@ -3507,13 +3517,22 @@ void call(client *c, int flags) {
+     * We leverage a static variable (prev_err_count) to retain
+     * the counter across nested function calls and avoid logging
+     * the same error twice. */
+    if ((server.stat_total_error_replies - prev_err_count)>0) {
```
style: space around '>'

madolson commented on src/server.h:
```diff
@@ -2132,7 +2138,9 @@ void updateDictResizePolicy(void);
 void resetCommandTableStats(void);
+void resetErrorTableStats(void);
 void adjustOpenFilesLimit(void);
+void incrementError(const char* fullerr, size_t namelen);
```
style: * typically sticks with the variable name in Redis.

filipecosta90 replied:
renamed to `incrementErrorCount`

filipecosta90 replied:
@madolson now fixed :)

madolson commented on src/server.c:
```diff
@@ -4904,14 +4941,34 @@ sds genRedisInfoString(const char *section) {
+        struct redisError *e;
+        while(raxNext(&ri)) {
+            e = (struct redisError *) ri.data;
+            if (!e->count) continue;
```
e->count is always at least 1 if it's present, so this check should be unnecessary.

madolson reviewed (dismissed):
My approval is for the new API syntax, I would still like to fix some of the minor comments.

filipecosta90 commented:
> My approval is for the new API syntax, I would still like to fix some of the minor comments.

@madolson @oranagra addressed the comments in the new commit.

yossigo reviewed (dismissed):
I'm not a great fan of the current approach of using arbitrary textual errors everywhere, and would be much happier if we had all possible error codes enumerated and mapped to error codes in one place (perhaps with the exception of modules). In that case we could also avoid the text manipulation and RAX which could end up be a bit more efficient.

That said, I doubt that performance is really an issue here (@filipecosta90 you can probably verify that in a breeze). Everything else is a matter of personal taste and I don't feel that strongly about it - so +1.

filipecosta90 commented:
> That said, I doubt that performance is really an issue here (@filipecosta90 you can probably verify that in a breeze). Everything else is a matter of personal taste and I don't feel that strongly about it - so +1.

@oranagra @yossigo @itamarhaber @madolson  I was getting the numbers for the standalone and cluster variations and noticed the following:
- performance is not affected here as you can check bellow
- one thing I believe we need to fix prior merge is that MOVED/CLUSTERDOWN/CROSSSLOT/... ( all error replies caused within `clusterRedirectClient()` are being accounted on the errorstats section, but not on the commandstats ( given we're not incrementing the rejected_calls there ). I believe we should... . If you agree tomorrow I'll fix it and include tests to verify it...

As a quick note @oranagra , to test the OSS Cluster variation I needed to update locally the branch to include the already merged changes by #8223.

----- 
## standalone numbers of an OOO error test

Used command and error replies:
```
$ redis-benchmark --threads 4 -h 10.3.0.110 -e -n 10000000 -t set
(...)
(...)
Error from server: OOM command not allowed when used memory > 'maxmemory'.
Error from server: OOM command not allowed when used memory > 'maxmemory'.
Error from server: OOM command not allowed when used memory > 'maxmemory'.
(...)
(...)
```

Without errorstats:
```
Summary:
  throughput summary: 204611.95 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.225     0.048     0.223     0.303     0.335     1.487
```

With errorstats:
```
Summary:
  throughput summary: 198503.30 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.232     0.048     0.231     0.311     0.343     3.319

$ $ redis-cli -h 10.3.0.110
10.3.0.110:6379> info errorstats
# Errorstats
errorstat_OOM:count=10000000
10.3.0.110:6379> info commandstats
# Commandstats
cmdstat_info:calls=2,usec=61,usec_per_call=30.50,rejected_calls=0,failed_calls=0
cmdstat_command:calls=2,usec=801,usec_per_call=400.50,rejected_calls=0,failed_calls=0
cmdstat_config:calls=5,usec=56,usec_per_call=11.20,rejected_calls=0,failed_calls=0
cmdstat_set:calls=348459,usec=106708,usec_per_call=0.31,rejected_calls=10000000,failed_calls=0
cmdstat_flushall:calls=1,usec=20,usec_per_call=20.00,rejected_calls=0,failed_calls=0
```


----- 
## OSS cluster numbers of an MOVED errors worst case scenario ( all moved )

Used command and error replies:
```
$ redis-benchmark --threads 4 -h 10.3.0.110 -p 30001 -e -n 10000000 -t set
(...)
(...)
Error from server: MOVED 13782 127.0.0.1:300038 (overall: 0.261).261)
Error from server: MOVED 13782 127.0.0.1:300030.407 (overall: 0.257))
Error from server: MOVED 13782 127.0.0.1:300032 (overall: 0.256).256)
(...)
(...)
```

Without errorstats:
```
Summary:
  throughput summary: 183874.23 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.252     0.048     0.255     0.327     0.359    18.207
```

With errorstats ( notice errorstats is populated with info but not commandstats ):
```
Summary:
  throughput summary: 184723.38 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.250     0.048     0.247     0.327     0.351    14.719

ubuntu@ip-10-3-0-195:~$ redis-cli -h 10.3.0.110 -p 30001 
10.3.0.110:30001> info errorstats
# Errorstats
errorstat_MOVED:count=10000000
10.3.0.110:30001> info commandstats
# Commandstats
cmdstat_psync:calls=1,usec=314,usec_per_call=314.00,rejected_calls=0,failed_calls=0
cmdstat_command:calls=1,usec=414,usec_per_call=414.00,rejected_calls=0,failed_calls=0
cmdstat_info:calls=4,usec=135,usec_per_call=33.75,rejected_calls=0,failed_calls=0
cmdstat_ping:calls=1,usec=0,usec_per_call=0.00,rejected_calls=0,failed_calls=0
cmdstat_cluster:calls=8,usec=1059,usec_per_call=132.38,rejected_calls=0,failed_calls=0
cmdstat_replconf:calls=204,usec=129,usec_per_call=0.63,rejected_calls=0,failed_calls=0
cmdstat_config:calls=2,usec=32,usec_per_call=16.00,rejected_calls=0,failed_calls=0

```

oranagra commented:
you meant #8226, yes, you should probably rebase this PR (can be a simple merge from unstable, since we're gonna squash this PR later)

regarding clusterRedirectClient I think it's a deeper problem, created: #8274 to discuss that.
meanwhile, i suggest you'll explicitly increment cmd->rejected_calls in processCommand after the call to clusterRedirectClient. if my PR will be merged, we'll get a merge conflict and resolve it properly.

sundb commented on src/server.c:
```diff
@@ -3674,6 +3668,7 @@ void call(client *c, int flags) {
 
     server.fixed_time_expire--;
     server.stat_numcommands++;
+    prev_err_count = server.stat_total_error_replies;
```
I have a question, why reset prev_err_count here.

oranagra replied:
this is why this variable exists:
```c
    /* Update failed command calls if required.
     * We leverage a static variable (prev_err_count) to retain
     * the counter across nested function calls and avoid logging
     * the same error twice. */
```

maybe it would have been nicer to move that reset line to right next to the place it is used (incrementing `failed_calls`)

enjoy-binbin commented:
We have encountered such a problem here. The user used such lua code:
```
return redis.error_reply(string.format('last:%s, now:%s....)
```

We can see `last:%s`, which will format the parameters. In errorstats, it will be used as an error code, causing more and more data saved in `info Errorstats`, and then when using INFO, the server will block since there are too many errorstats.
```
#Errorstats
errorstat_ERR:count=27
errorstat_last_1709544349,:count=1
errorstat_last_1709544354,:count=1
errorstat_last_1709544359,:count=1
errorstat_last_1709544360,:count=1
errorstat_last_1709544371,:count=1
errorstat_last_1709544385,:count=1
errorstat_last_1709544390,:count=1
errorstat_last_1709544395,:count=1
errorstat_last_1709544396,:count=1
errorstat_last_1709544408,:count=1
errorstat_last_1709544413,:count=2
errorstat_last_1709544414,:count=2
errorstat_last_1709544432,:count=1
errorstat_last_1709544442,:count=1
errorstat_last_1709544450,:count=1
```

Is there any way we can control this incorrect usage?

oranagra commented:
i'm not sure there's anything we can do.
the API is clearly documented that the first word is the error code, and i don't think we have any way to distinguish between a misuse and a valid one.
the only thing we can do is maybe put some maximum cap on the extent of damage it does, and disable the error stats mechanism completely if there are more than say 1000 different errors.
i.e. print them to the log file, empty the dict, and set a flag to avoid populating it from now on.

enjoy-binbin commented:
> the only thing we can do is maybe put some maximum cap on the extent of damage it does, and disable the error stats mechanism completely if there are more than say 1000 different errors.

yean, the only way i thought of is the same. in info errorstats, use raxSize and only part of the data is output to break out of the loop. Note that it may also potentially consume some memory (if it has been accumulated)

Or set a parameter to limit the number of Errorstats we save

oranagra commented:
if we do that, i'd rather stop collecting them, release the memory and permanently disable error stats. not just filter them at reporting time.

enjoy-binbin commented:
yes, i think so too, just.. i can’t describe it very well.

ok, what i think is similar to only saving up to 1000 errorstats, and deleting some of them when it exceeds, similar to our earliest repl_scriptcache_fifo, something like that

but just some random ideas, just to throw the discussion out there first

oranagra commented:
i don't think we'd want to delete some. i think we can detect a misuse and completely disable this feature preventing the damage it can cause, and losing the functionality.
to help users figure out what happened, i'd print the existing values to the log file, add a single entry to replace them that mentioned it's disabled due to misuse, and prevent any further accumulation.
